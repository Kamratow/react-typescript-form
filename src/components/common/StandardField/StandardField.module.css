.StandardFieldWrapper {
    margin: 0;
    box-sizing: border-box;
    padding: 5px;
}

.StandardFieldLabel {
    font-size: 20px;
    color: #000;
    font-weight: 600;
    display: block;
    margin-bottom: 8px;
}

.StandardFieldInput {
    font-size: 20px;
    padding: 10px;
    border-radius: 5px;
    width: 100%;
    color: #000;
    border: 1px solid #000;
    box-sizing: border-box;
    margin-bottom: 24px;
}

.StandardFieldInput:focus {
    outline: none;
    box-shadow: 0 0 6pt #008000;
    border-color: #008000;
}

.StandardFieldInputError {
    border-color: rgb(201, 22, 22);
    margin-bottom: 0;
}

.StandardFieldErrorMessage {
    font-size: 16px;
    color: rgb(201, 22, 22);
    margin-top: 3px;
    margin-bottom: 0;
}
