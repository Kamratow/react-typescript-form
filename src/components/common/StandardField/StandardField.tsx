import React, { useState } from "react";

import styles from "./StandardField.module.css";
import { IFormInput } from "../../../models/formInput.model";

export interface IStandardFieldProps extends IFormInput {
    changeHandler: (newValue: string) => void;
}

const StandardField: React.FunctionComponent<IStandardFieldProps> = ({
    id,
    label,
    type,
    value,
    changeHandler,
    errorMessage,
}) => {
    const [isTouched, setIsTouched] = useState<boolean>(false);

    const handleOnChange = (event: React.FormEvent<HTMLInputElement>) => {
        setIsTouched(true);

        changeHandler(event.currentTarget.value);
    };

    const getInputClasses = (): string => {
        const inputClasses: string[] = [];
        inputClasses.push(styles.StandardFieldInput);

        if (errorMessage !== "") {
            inputClasses.push(styles.StandardFieldInputError);
        }

        return inputClasses.join(" ");
    };

    return (
        <div className={styles.StandardFieldWrapper}>
            <label htmlFor={id} className={styles.StandardFieldLabel}>
                {label}
            </label>
            <input
                id={id}
                type={type}
                value={value}
                className={getInputClasses()}
                onChange={handleOnChange}
            />
            {errorMessage !== "" && isTouched && (
                <p
                    id={id + "ErrorMessage"}
                    className={styles.StandardFieldErrorMessage}
                >
                    {errorMessage}
                </p>
            )}
        </div>
    );
};

export default StandardField;
