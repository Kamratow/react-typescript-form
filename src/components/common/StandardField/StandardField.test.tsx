import React from "react";
import { shallow } from "enzyme";

import StandardField, { IStandardFieldProps } from "./StandardField";

describe("StandardField", () => {
    let defaultProps: IStandardFieldProps;
    let changeHandlerMock: () => {};

    beforeEach(() => {
        changeHandlerMock = jest.fn();
        defaultProps = {
            changeHandler: changeHandlerMock,
            errorMessage: "",
            id: "testId",
            label: "testLabel",
            type: "text",
            value: "test value",
        };
    });

    it("should render properly with default props", () => {
        const wrapper = shallow(<StandardField {...defaultProps} />);

        expect(wrapper).toMatchSnapshot();
    });

    it("should properly handle change event", () => {
        const wrapper = shallow(<StandardField {...defaultProps} />);
        const input = wrapper.find("#testId").at(0);
        const mockedChangeEvent = {
            currentTarget: {
                value: "changed value",
            },
        };

        expect(changeHandlerMock).toHaveBeenCalledTimes(0);

        input.simulate("change", mockedChangeEvent);

        expect(changeHandlerMock).toHaveBeenCalledTimes(1);
        expect(changeHandlerMock).toHaveBeenCalledWith("changed value");
    });

    it("should not render error if input has not been touched", () => {
        const propsToRender: IStandardFieldProps = {
            ...defaultProps,
            errorMessage: "test error message",
        };

        const wrapper = shallow(<StandardField {...propsToRender} />);

        expect(wrapper).toMatchSnapshot();
    });

    it("should render error if input has been touched", () => {
        const propsToRender: IStandardFieldProps = {
            ...defaultProps,
            errorMessage: "test error message",
        };
        const wrapper = shallow(<StandardField {...propsToRender} />);
        const input = wrapper.find("#testId").at(0);
        const mockedChangeEvent = {
            currentTarget: {
                value: "changed value",
            },
        };

        input.simulate("change", mockedChangeEvent);

        expect(wrapper).toMatchSnapshot();
    });
});
