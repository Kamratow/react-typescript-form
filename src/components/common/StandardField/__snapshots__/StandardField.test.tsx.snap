// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StandardField should not render error if input has not been touched 1`] = `
<div
  className="StandardFieldWrapper"
>
  <label
    className="StandardFieldLabel"
    htmlFor="testId"
  >
    testLabel
  </label>
  <input
    className="StandardFieldInput StandardFieldInputError"
    id="testId"
    onChange={[Function]}
    type="text"
    value="test value"
  />
</div>
`;

exports[`StandardField should render error if input has been touched 1`] = `
<div
  className="StandardFieldWrapper"
>
  <label
    className="StandardFieldLabel"
    htmlFor="testId"
  >
    testLabel
  </label>
  <input
    className="StandardFieldInput StandardFieldInputError"
    id="testId"
    onChange={[Function]}
    type="text"
    value="test value"
  />
  <p
    className="StandardFieldErrorMessage"
    id="testIdErrorMessage"
  >
    test error message
  </p>
</div>
`;

exports[`StandardField should render properly with default props 1`] = `
<div
  className="StandardFieldWrapper"
>
  <label
    className="StandardFieldLabel"
    htmlFor="testId"
  >
    testLabel
  </label>
  <input
    className="StandardFieldInput"
    id="testId"
    onChange={[Function]}
    type="text"
    value="test value"
  />
</div>
`;
